res/controllers/Denon-Prime-4-scripts.js

```diff
@@ -0,0 +1,869 @@
+        input: function(channel, control, value, status, _group) {
+            if (!this.isPress(channel, control, value, status)) {
+                return;
+            }
+            theDeck.jogWheel.vinylMode = !theDeck.jogWheel.vinylMode;
+            this.output(theDeck.jogWheel.vinylMode);
+        },
+        trigger: function() {
+            this.output(theDeck.jogWheel.vinylMode);
+        },
+    });
+
+    // Jog Wheel
+    this.jogWheel = new components.JogWheelBasic({
+        deck: script.deckFromGroup(this.currentDeck),
+        wheelResolution: 1000,
+        alpha: 1/8,
+        beta: 1/8/32,
+        rpm: 33 + 1/3,
+        // Instead of relative movements between this and the last position,
+        // the controller reports the absolute position of the wheel with
+        // 14-bit precision. Because of that, we need to reconstruct the value
+        // and then transform it into the relative directions expected by Mixxx.
+        inputWheelMSB: function(_channel, _control, value, _status, _group) {
+            this.wheelMSB = value;
+        },
+        inputWheelLSB: function(channel, control, value, status, group) {
+            this.inputWheel(channel, control, (this.wheelMSB << 7) + value, status, group);
+        },
+        previousPosition: null,
+        wrappingValue: Math.pow(2, 14),
+        relativeFromAbsolute: function(value) {
+            // The first value of the controller will probably be random
+            // and thus we just have to swallow it until we have the second value
+            // to find the difference
+            if (this.previousPosition === null) {
+                this.previousPosition = value;
+                return 0;
+            }
+            // This finds the shortest distance between the current value
+            // and the last one, and preserves the orientation
+            const delta = value - this.previousPosition;
+            let remainder = ((delta % this.wrappingValue) + this.wrappingValue) % this.wrappingValue;
+            if (remainder * 2 > this.wrappingValue) {
+                remainder -= this.wrappingValue;
+            }
+            this.previousPosition = value;
+            return remainder;
+        },
+        jogScale: function(val) {
+            // wheelSensitivity is user-configurable, see top of file.
+            return val * wheelSensitivity;
+        },
+        inputWheel: function(channel, control, value, _status, _group) {
+            value = this.relativeFromAbsolute(value);
+            if (engine.isScratching(this.deck)) {
+                engine.scratchTick(this.deck, value);
+            } else {
+                this.inSetValue(this.jogScale(value));
+            }
+        },
+    });
+
+    // Jog Wheel LED
+    this.jogWheelLed = new components.Component({
+        midi: [0x90 + midiChannel, 0x21],
+        trigger: function() {
+            midi.sendShortMsg(0x90 + midiChannel, 0x21, colDeck[midiChannel - 4]);
```

Comment:
Figured it out. `midiChannel` is only ever 4 or 5, so it would send the 1/2 color even when switching to 3/4. Also, we should really rename `deckNumbers` to singular `deckNumber`.

```suggestion
        on: colDeck[deckNumbers - 1],
        trigger: function() {
            this.send(this.on);
```

Reply:
Thank you. I'll rename `deckNumbers` shortly. The jog wheel LEDs work as expected now.